### PR TITLE
Use decompositions for some fallbacks introduced in #94039

### PR DESCRIPTION
**Pull request by fdrocha**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94206

In some cases, implements required inductor primitives.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: torch/_inductor/decomposition.py [hunk 1]
```diff
@@ -22,13 +22,53 @@
...
+        aten.binary_cross_entropy_backward,
+        aten.celu,
+        aten.diag_embed,
+        aten.diagonal,
+        aten.frac,
+        aten.hardshrink,
+        aten.heaviside,
+        aten.huber_loss,
+        aten.isneginf,
+        aten.isposinf,
```
Comment (ngimel) [inline torch/_inductor/decomposition.py:45]:
```suggestion
```

File: torch/_inductor/decomposition.py [hunk 2]
```diff
@@ -22,13 +22,52 @@
...
         aten.std,
         aten.std_mean,
-        aten._to_copy,
         aten.triu_indices,
         aten.tril_indices,
         aten.sqrt_,
         aten.lcm,
         aten.clamp_min_,
         aten.sin_,
+        aten.addcdiv,
```
Comment (ngimel) [inline torch/_inductor/decomposition.py:31]: most (if not all) of these additions should go to core_aten_decompositions in _decomp/__init__.py, cc @SherlockNoMad

File: torch/_inductor/decomposition.py [hunk 3]
```diff
@@ -22,13 +22,52 @@
...
+        aten.celu,
+        aten.diag_embed,
+        aten.diagonal,
+        aten.frac,
+        aten.hardshrink,
+        aten.heaviside,
+        aten.huber_loss,
+        aten.isneginf,
+        aten.isposinf,
+        aten.logical_xor,
```
Comment (ngimel) [inline torch/_inductor/decomposition.py:46]: imo better to have lowering for this (together with other logical_* ops)

**Comment by ngimel** [comment]
@fdrocha can you please rebase? Automatic rebase doesn't work

File: torch/_inductor/decomposition.py (hunk 3 above)
Comment (fdrocha) [reply torch/_inductor/decomposition.py:46]: Sure, will add those

File: torch/_inductor/decomposition.py (hunk 2 above)
Comment (fdrocha) [reply torch/_inductor/decomposition.py:31]: Right, I wasn't sure what was the reasoning for something going here vs `core_aten_decompositions`. What's the idea there?

**Comment by fdrocha** [comment]
> @fdrocha can you please rebase? Automatic rebase doesn't work

Yes, will do and address your comments before merging

File: torch/_inductor/decomposition.py (hunk 3 above)
Comment (ngimel) [reply torch/_inductor/decomposition.py:46]: other `logical_*` already exist in lowering, so `logical_xor` better fits with existing `logical_and` and friends,  and `@register_pointwise` has `convert_to_bool` arg which is pretty much what decomp is doing.

File: torch/_inductor/decomposition.py (hunk 2 above)
Comment (ngimel) [reply torch/_inductor/decomposition.py:31]: `core_aten_decompositions` is for decompositions that do aten->aten (don't lower to prims). They can be used by backends that support aten ops only, no prims.

File: torch/_decomp/__init__.py [hunk 4]
```diff
@@ -176,116 +176,152 @@ def core_aten_decompositions() -> Dict[OpOverload, Callable]:
...
             aten.index_copy_,
             aten.index_fill,
             aten.index_fill_,
-            aten.isposinf,
+            aten.index_select,
+            aten.isneginf,
             aten.isneginf,
+            aten.isposinf,
+            aten.isposinf,
+            aten.isposinf,
```
Comment (peterbell10) [inline torch/_decomp/__init__.py:237]: What's going on here, is the duplication intentional?

Comment (fdrocha) [reply torch/_decomp/__init__.py:237]: thanks, good catch, not sure how that happened

File: torch/_inductor/lowering.py [hunk 5]
```diff
@@ -3871,6 +3801,28 @@ def sum_(x, axis=None, keepdims=False, *, dtype=None):
         override_return_dtype=torch.bool,
     )
 )
+logical_xor = register_pointwise(
+    aten.logical_xor,
```
Comment (ngimel) [inline torch/_inductor/lowering.py:3805]: nit, can be done in follow up, if you send `name=bitwise_xor` arg chances are you won't have to implement logical_xor op, but if it doesn't work, don't bother

File: torch/_inductor/codegen/cpp.py [hunk 6]
```diff
@@ -706,6 +813,10 @@ def logical_and(a, b):
     def logical_or(a, b):
         return f"{a} || {b}"
 
+    @staticmethod
+    def logical_xor(a, b):
```
Comment (ngimel) [inline torch/_inductor/codegen/cpp.py:817]: unless you manage to get rid of it, it can go into `common.py` as it's the same between cpu and gpu?

**Review by ngimel — APPROVED** [review]

File: torch/_inductor/codegen/cpp.py (hunk 6 above)
Comment (ngimel) [reply torch/_inductor/codegen/cpp.py:817]: https://github.com/pytorch/pytorch/blob/master/torch/_inductor/codegen/common.py#L144

File: torch/_inductor/lowering.py (hunk 5 above)
Comment (fdrocha) [reply torch/_inductor/lowering.py:3805]: This works nicely, thanks!
You can actually do same thing for logical_and and logical_or... But maybe that breaks some desired short-circuiting behavior, so I didn't do it.

**Pull request closed without merging**